File: 2020/03/aoc03.erl
-module(aoc03).
-export([run1/0, run2/0]).

read(File) -> 
    {ok, Bin} = file:read_file(File),
    string:lexemes(binary_to_list(Bin), "\n").

check(Pos, Line) -> 
    case lists:nth(Pos + 1, Line) of 
        $# -> 1;
        _ -> 0
    end.

trees({R, L}, Items) -> 
    Indexed = lists:zip(lists:seq(0, length(Items) - 1), Items),
    Included = lists:filter(fun ({Ix, _}) -> (Ix rem L) == 0 end, Indexed),
    Values = lists:map(fun ({Ix, Line}) -> check((R * Ix) rem (length(Line)), Line) end, Included),
    lists:sum(Values).

run(Configs) -> 
    Input = read("input.txt"),
    Results = lists:map(fun (C) -> trees(C, Input) end, Configs),
    lists:foldl(fun (X, Prod) -> X * Prod end, 1, Results).

run1() -> 
    run([{3, 1}]).

run2() -> 
    run([{1, 1}, {3, 1}, {5, 1}, {7, 1}, {1, 2} ]).

File: 2020/03/aoc03.fs
open System.IO

let trees (lines : string array) (right : int, left : int) : int64 =
    lines
    |> Array.mapi (fun i line -> (i, line))
    |> Array.filter (fun (i, _) -> i % left = 0)
    |> Array.map (fun (i, line) -> if '#' = line.[(right * i) % line.Length] then 1L else 0L)
    |> Array.sum

let run (lines : string array) (configs : (int * int) list) =
    configs
    |> List.map (trees lines)
    |> List.reduce (*)
    |> printfn "%d"

[<EntryPoint>]
let main argv =
    let lines = File.ReadAllLines argv.[0]
    [ (3, 1) ] |> run lines
    [ (1, 1); (3, 1); (5, 1); (7, 1); (1, 2) ] |> run lines
    0 